Added log exporter tool for uploading k8s cluster logs directly to GCS

shyamjvs opened the pull request:
This tool would help save time in our tests (currently ~4 hours for 2000 node tests) by making nodes in the cluster upload logs to GCS directly rather than first copying them over to the Jenkins node (through scp) and then uploading to GCS from there. For more info, read the README.md file. This PR would be followed up with another one to make this the logging option for our tests.
Thanks @crassirostris for discussing how this approach is better than existing open source logging tools.

cc @kubernetes/sig-scalability-misc @kubernetes/sig-testing-misc @wojtek-t @gmarek @fejta

fejta commented on logexp/logexp.sh:
```diff
@@ -0,0 +1,148 @@
+set -o nounset
+set -o pipefail
+
+# The following env vars need to be defined before calling this script:
```
I would prefer we pass in arguments rather than rely on environment variables. Any chance on making this change?

Also thoughts on writing this in a language other than bash (python or golang?)

wojtek-t replied:
+1 for golang with flags

shyamjvs replied:
Is introducing flags the only reason for suggesting golang? Because I can move it to flags in bash too, using getopts and its in POSIX standard ([ref](http://mywiki.wooledge.org/BashFAQ/035))
Also most of the code is simply just scripting (like gcloud, gsutil, journalctl and other calls), not sure if a full-blown programming language would be convenient. WDYT?

wojtek-t replied:
I don't agree with you. The script has 150 lines and there are:
- 1 gcloud
- 1 gsutil
- 3 journalctl
The rest is the normal code.
And i think we are generally trying to reduce amount of bash in our codebase.

fejta replied:
The flag and golang suggestion are separate.

Flags make it easier to trace dependencies and predict the behavior of the program based on how the program is launched vs global environment state where it was launched.

Golang makes it easier to predict the impact of changes. It is also easier to write unit tests (please consider writing some unit tests to cover interesting and/or non-trivial scenarios):

* program exits if the service account creds are not specified
* we add the right files based on the provider
* we do not crash if journalctl fails
* we retry gsutil uploads that flake
* we detect systemd vs supervisord correctly
* etc

ixdy commented on logexp/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+# This file builds an image for the log exporter tool. For more info,
+# have a look at the tool's README.md file.
+
+FROM gcr.io/google-containers/kubekins-test:v20160822
```
kubekins-test has a bunch of stuff you don't need, I suspect. maybe just use `python-slim` (which removes the need to install python), or use `gcr.io/google-containers/debian-base-amd64:0.1` (along with `clean-install python`)

shyamjvs replied:
Done. I was desperately searching for one such small image that has python, thanks a lot.
After trying alpine, python:slim and debian, it seemed most easiest/smallest to make it work on alpine, so sticked to it.

ixdy commented on logexp/Makefile:
```diff
@@ -0,0 +1,28 @@
+all: build
+
+build: logexp.sh
+	docker build -t $(IMG):$(TAG) .
```
add `--pull`

shyamjvs replied:
Done.

ixdy commented on logexp/Makefile:
```diff
@@ -0,0 +1,28 @@
+
+build: logexp.sh
+	docker build -t $(IMG):$(TAG) .
+	docker tag $(IMG):$(TAG) $(IMG):latest
```
we generally avoid using `:latest` tags, I think.

shyamjvs replied:
This is a convention I followed from https://github.com/kubernetes/test-infra/blob/master/jenkins/e2e-image/Makefile#L26
And usually it is suggested to not have prod containers run against latest tag. For a volatile component in test, especially like this one, which is tagged with a computed hash and timestamp, the latest tag might be convenient. WDYT?

fejta replied:
I agree that having a latest tag is useful for a variety of testing the image purposes.

fejta commented on logexp/Makefile:
```diff
@@ -0,0 +1,28 @@
+# limitations under the License.
+
+IMG = gcr.io/google-containers/logexp
+TAG = v0.1.0
```
Consider this pattern to automatically generate the tag name that includes the hash used to create it.
https://github.com/kubernetes/test-infra/blob/master/jenkins/e2e-image/Makefile#L16

shyamjvs replied:
Had that originally, but dropped it assuming that this would be a more stable image (not built regularly like the kubekins image probably) and could be used even outside test-infra.
Changed it now though, let's stick to that pattern. Thanks!

shyamjvs commented:
Sorry I got busy with something else in the meanwhile. Moved the script to golang, looks much cleaner now.
And using job with hard inter-pod anti-affinity to make sure exactly one pod of the exporter runs per node, until completion.
I've tested that it works on gce by running the job on my own 3-node cluster, spawned with kube-up.

@fejta @wojtek-t PTAL now.

crassirostris commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+
+	// Node-type specific logfiles.
+	masterLogs = []string{"kube-apiserver", "kube-scheduler", "rescheduler", "kube-controller-manager", "etcd", "etcd-events", "glbc", "cluster-autoscaler", "kube-addon-manager", "fluentd"}
+	nodeLogs = []string{"kube-proxy", "node-problem-detector", "fluentd"}
```
What if this list changes? Shouldn't it be configured also?

fejta replied:
I'd prefer not making this configurable at first. If the list changes we'll probably want to ship a new version of this binary with the updated list.

fejta commented on logexp/Makefile:
```diff
@@ -0,0 +1,34 @@
+all: build
+
+cmd/logexp: cmd/main.go
+	CGO_ENABLED=0 go build -o cmd/logexp k8s.io/test-infra/logexp/cmd
```
If you compare with https://github.com/kubernetes/test-infra/blob/master/jenkins/e2e-image/Makefile#L27 you'll notice that we delete the compiled binary after the build step. We do this to reduce the chance of pushing a commit with bad data. Consider copying this pattern?

shyamjvs replied:
Done.

fejta commented on logexp/README.md:
```diff
@@ -0,0 +1,39 @@
+# LogExp
+
+LogExp is a tool that would run post-test on our kubernetes test clusters.
```
nit: s/would run/runs/

shyamjvs replied:
Done.

fejta commented on logexp/README.md:
```diff
@@ -0,0 +1,39 @@
+node type (master/node), cloud provider, and the node's system services),and
+then actually exports them to the GCS path provided to it.
+
+The following environment variables need to be supplied to the tool:
```
I believe these are now flags?

shyamjvs replied:
Done. Thanks for noticing!

fejta commented on logexp/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## How to run the tool?
+
+Before running the tool on any node, setup gcloud service account credentials through a secret:
```
nit: create a secret with service account credentials?

shyamjvs replied:
Done.

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+	}
+	glog.Infof("Finished uploading logs successfully")
+
+	glog.Infof("Entering sleep for a duration of %v seconds", sleepDuration)
```
Maybe guard this inside `sleepDuration > 0`?

shyamjvs replied:
Done.

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+  		}
+  		break
+  	}
+  	if uploadAttempt == 3 {
```
Might be worth saving the first and/or last error and including this in the returned error?

shyamjvs replied:
Included the last one in the returned error (as technically the final attempt failing is what should be considered the error). Is it fine?

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+  			glog.Errorf("Attempt %v to upload to GCS failed: %v", uploadAttempt, err)
+  			continue
+  		}
+  		break
```
return nil? If so then you can limit uploadAttempt scope to the for loop?

shyamjvs replied:
Cool suggestion! Done.

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+
+	// Upload the files with compression and parallelism (speeds up on good network)
+  	// along with setting their ACL to make them publicly readable.
+  	gcsLogPath = filepath.Join(gcsArtifactsPath, nodeName)
```
Will this cause weird behavior on windows?
Consider using net/url?

Maybe not necessary

shyamjvs replied:
Well it seems like it should be fine for windows too, as the doc says "On Windows, the result is a UNC path if and only if the first path element is a UNC path." (Ref: https://golang.org/pkg/path/filepath/#Join).
In fact given that this would be run inside a linux container, shouldn't be much of an issue. Moreover, there are many linux'ish assumptions throughout the code ("/var/log", "/bin/sh", "sudo journalctl", etc)

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+		glog.Infof("Journalctl found on host. Collecting systemd logs")
+		createSystemdLogfiles(logDir, isMaster)
+	} else {
+		glog.Infof("Journalctl not found on host (%v). Collecting supervisord logs instead", err)
```
Any interest in distinguishing IsNotExist from other types of errors?

https://golang.org/pkg/os/#IsNotExist

shyamjvs replied:
Good catch. Not particularly interested in IsNotError, as we'd want any error to lead there. Changed it now. Is this fine?

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+	} else {
+		journalCmd = fmt.Sprintf("sudo journalctl --output=%v -u %v.service", outputMode, service)
+	}
+	cmd := exec.Command("/bin/sh", "-c", journalCmd)
```
Not sure I understand why we need/want the /bin/sh prefix?

shyamjvs replied:
I'll try to fix this by getting rid of the sudo

wojtek-t replied:
+1 to what @fejta wrote

shyamjvs replied:
This one is a tricky issue and I failed to find any solution so far. Even the current command (with or without sudo) fails to run. The problem is with using ubuntu base image. As pointed out by @Random-Liu (who also previously tried using journalctl for npd):
The journald in ubuntu compiled with incompatible compile configuration with GCI.
```
Journal file /var/log/journal/0e429019876a34f9c55ec02a99498747/system.journal uses an unsupported feature, ignoring file.
-- No entries --
```

In the case of NPD however, we could get away by using fedora, for which journalctl works fine. However for our use, we also need to install gcloud and it fails on fedora due to python issue.
I'm trying to make gcloud work on fedora currently (with no success so far). Is anyone aware of a base image that could be useful here?

shyamjvs replied:
Ok. I finally managed to make it work on fedora's latest image (v25) by installing 'which' (required by google-cloud-sdk/install.sh) and 'python2.7' (as gcloud needs version 2.7 - https://cloud.google.com/sdk/downloads#system_requirements). I then tested the image on a GCI cluster (having systemd services) and the job uploaded the logs successfully.

shyamjvs replied:
Also got rid of /bin/sh and now we're calling journalctl directly.

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+	for _, logfile := range logfiles {
+		logfileFullPath := filepath.Join(localLogPath, logfile + ".log*")  // Append .log* to copy rotated logs too.
+		cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("cp %v %v", logfileFullPath, logDir))
+		if _, err := cmd.Output(); err != nil {
```
Given that we are ignoring the Output() could we just call Wait() instead?

shyamjvs replied:
Seems like every call to Wait should be preceded by a call to Start. However, [Run](https://golang.org/pkg/os/exec/#Cmd.Run) seems to be the option we are looking for. Changed it at all relevant places.

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,257 @@
+var (
+	// Source and destination logs directory.
+	localLogPath = "/var/log"
+	gcsLogPath = ""  // Computed as "gcsArtifactsPath/nodeName"
```
Is there a reason this needs to be global?

shyamjvs replied:
No good reason, removed it now. Just wanted to make all config-related variables visible at the top.

fejta commented on logexp/cluster/job.yaml:
```diff
@@ -0,0 +1,70 @@
+        command:
+        - logexp
+        - --node-name=$(NODE_NAME)
+        - --kubernetes-provider={{.KubernetesProvider}}
```
This is pretty nifty, can you link me to a doc that shows how this works? Where does this value come from??

shyamjvs replied:
This is a standard for templating in golang. Ref: https://golang.org/pkg/text/template/#example_Template

fejta replied:
Hmm, I still don't understand where `KubernetesProvider` and `GcloudAuthFilePath` come from?

shyamjvs replied:
These are template variables that need to be filled in at the place where we are going to create the job (which would be somewhere near function [DumpClusterLogs](https://github.com/kubernetes/test-infra/blob/master/kubetest/e2e.go#L391) in e2e.go. This also calls for making these argument values available inside e2e.go (some of which we already have))

fejta commented on logexp/Makefile:
```diff
@@ -0,0 +1,34 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+PROJECT ?= google-containers
```
When does this change?

Also consider using k8s-testimages

shyamjvs replied:
Oops.. Sorry for the blooper, fixed it. I didn't want to push any image to k8s-testimages before we agree upon this. So:
- Would I be able to push to it (given that I have permission to push to google-containers) or would you( /testinfra) do the pushing part?
- Given that this (in part) does the job of cluster/log-dump.sh, maybe this tool should be available under the more general project viz. google-containers?
- We might want to decide upon the tag based on the project(?) Like v0.1.0 or something if we want to use google-containers, or git describe --tags based tag for k8s-testimages?

wojtek-t commented on logexp/Makefile:
```diff
@@ -0,0 +1,35 @@
+
+push: build
+	gcloud docker -- push $(IMG):$(TAG)
+	gcloud docker -- push $(IMG):latest
```
Why do we push "latest" too? I think we're generally trying to avoid using "latest" tag.
@fejta - is this some common pattern?

fejta replied:
Yes, we use the immutable tag on the previous line for things that need a specific version. 

The latest tag is useful for testing new images before using them in production.

This is the pattern we use in this repo.

wojtek-t commented on logexp/README.md:
```diff
@@ -0,0 +1,40 @@
+# LogExp
```
We should consider renaming it. "LogExp" for me is associating with "logarithm" and "exponential" (which has nothing to do with it).
Maybe just "LogExporter" ?

shyamjvs replied:
I thought of giving a short/crisp name originally. But LogExporter sounds better. Changed it.

wojtek-t commented on logexp/README.md:
```diff
@@ -0,0 +1,40 @@
+then actually exports them to the GCS path provided to it.
+
+The following flags need to be supplied to the tool:
+* --node-name (Name of the node this log exporter is running on)
```
Explaining what "node name" is seems useless to me.

shyamjvs replied:
Explained it just to be similar with other flags. Removed the whole para anyway from the Readme now.

wojtek-t commented on logexp/README.md:
```diff
@@ -0,0 +1,40 @@
+
+The following flags need to be supplied to the tool:
+* --node-name (Name of the node this log exporter is running on)
+* --gcs-artifacts-path (Path to the GCS artifacts dir under which to upload logs, for eg: gs://my-logs-bucket/logs)
```
s/artifacts//

shyamjvs replied:
Done.

wojtek-t commented on logexp/README.md:
```diff
@@ -0,0 +1,40 @@
+The following flags need to be supplied to the tool:
+* --node-name (Name of the node this log exporter is running on)
+* --gcs-artifacts-path (Path to the GCS artifacts dir under which to upload logs, for eg: gs://my-logs-bucket/logs)
+* --kubernetes-provider (Cloud provider for this node (gce/gke/aws/kubemark/..))
```
aren't we using "--provider" in different places of the codebase?

shyamjvs replied:
Seems like we are using --cloud-provider (Ref: https://github.com/kubernetes/kubernetes/blob/master/hack/verify-flags/known-flags.txt#L100)
Changed it now. Is it fine?

wojtek-t commented on logexp/README.md:
```diff
@@ -0,0 +1,40 @@
+node type (master/node), cloud provider, and the node's system services),and
+then actually exports them to the GCS path provided to it.
+
+The following flags need to be supplied to the tool:
```
I don't think we should explain flags of the tool in Readme - if you need to understand them the binary itself should expose help (and if you're using standard flag package it does).

shyamjvs replied:
True. Help exposes them anyway, so removed it.

wojtek-t commented on logexp/README.md:
```diff
@@ -0,0 +1,40 @@
+1. Prepare service-account.json file (that should have write access to specified GCS path)
+2. `kubectl create secret generic logexp-service-account --from-file=service-account.json=/path/to/service-account.json`
+
+To run the tool as a pod on a single node:
```
Do we have usecase of it? I think we won't be using it, and in that case just remove this.

shyamjvs replied:
I kept it just for quick testing on a single node rather than doing on whole cluster. Also it'll be handy if someone is interested in just a single node's logs instead of the entire cluster on which he might be running an e2e test. Is that a reasonable use or should I remove it?

wojtek-t commented on logexp/cluster/job.yaml:
```diff
@@ -0,0 +1,72 @@
+# Template job config for running the log exporter on a whole kubernetes cluster.
+# We ensure that exactly one logexp pod runs per node using hard inter-pod anti-affinity.
+# Steps:
```
This is already in readme - please don't duplicate it here.

shyamjvs replied:
Removed it.

wojtek-t commented on logexp/cluster/pod.yaml:
```diff
@@ -0,0 +1,54 @@
+# Template pod config for running the log exporter pod on a kubernetes node/master.
```
Are we going to use it somewhere? If not - just remove this file.

shyamjvs replied:
Discussed in [this](https://github.com/kubernetes/test-infra/pull/2310#discussion_r109102384) comment.

wojtek-t commented on logexp/cmd/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+
+# Setup the log exporter script.
+ADD ["logexp", \
+     "/workspace/"]
```
Single line please

shyamjvs replied:
Done.

wojtek-t commented on logexp/cmd/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+     "/workspace/"]
+WORKDIR "/workspace"
+
+ENTRYPOINT ["/workspace/logexp"]
```
I'm slightly against having it - I really prefer if the image is not doing anything by default and the user is supposed to specify a commandline - as it has better control over it then.

fejta replied:
I prefer we enforce the entrypoint.

The image has a single purpose and should know what to do. The name of the binary, its path within the container, etc should all be something the image concerns itself with, not the caller of the image.

I do not see the need of exposing this complexity to the user nor their need for flexibility. They can derive their own image from this one if that is needed.

Otherwise all a user should have to do is pass in the well known flags which control the image's behavior.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+limitations under the License.
+*/
+
+// This script when run on a node (/master) in a kuberentes cluster would
```
We already have a readme - I don't think this comment is useful.

shyamjvs replied:
Removed it.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+)
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&nodeName, "node-name", "", "Name of the node this log exporter is running on")
```
In case of those simple binaries, we are using a different patter for flags.
So instead of declaring variables and then assigning flags to them we do, e.g.:
```
var (
  nodeName = flag.String("node-name", "", "...")
  ...
)
```

See e.g. https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/attachlease/attachlease.go#L29 as an example.

shyamjvs replied:
Changed it. Looks simpler now. Thanks!

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+// Create logfiles for systemd services in outputDir, based on whether the node isMaster.
+func createSystemdLogfiles(outputDir string, isMaster bool) {
+	services := systemdServices
+	if ! isMaster {
```
s/! /!/

shyamjvs replied:
Done.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	case "aws":
+		logfiles = append(logfiles, awsLogs...)
+	default:
+		glog.Warningf("Unknown cloud provider '%v' provided, skipping any provider specific logs", kubernetesProvider)
```
Errorf

shyamjvs replied:
Done.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	var isMaster bool
+
+	// Select node-type (master/node) specific logs.
+	if _, err := os.Stat(localLogPath + "/kube-apiserver.log"); err == nil {
```
This is hacky - if we e.g. decide to change the log file path, then this will stop working.
Also, because of e.g. some problems with apiserver, its logs may not be there, and we still want to grab all other logs.

shyamjvs replied:
Removed it. Infact removed all the logging code specific to master as currently this doesn't anyway work for master (due to some gcloud auth issue).
And in fact the main chunk of the time we are trying to optimize is that due to copying of logs from nodes. Master logs copying can still continue to be done by scp'ing through jenkins.
Also, this way we would not have to complicate our existing setup for kubemark master too.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	}
+}
+
+// Effeciently upload all logfiles inside logDir to GCS.
```
Method name is self-explaining, this comment is useless in my opinion.

shyamjvs replied:
Removed it. Sorry for the overwhelming usage of comments :)

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+		logfiles = append(logfiles, nodeLogs...)
+	}
+
+	// Select cloud-provider specific logs.
```
Remove this comment.

shyamjvs replied:
Done.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	case "gce", "gke":
+		logfiles = append(logfiles, gceLogs...)
+	case "kubemark":
+		logfiles = append(logfiles, gceLogs...)  // TODO(shyamjvs): Make this variable wrt kubemark's real provider.
```
I don't understand this TODO.

BTW - move it to previous line.

shyamjvs replied:
Moved it. What I meant by that TODO was, inside kubemark we again need to decide whether to pick gceLogs / awesLogs / etc based on the real cloud provider.
It doesn't hurt for now, as we run kubemark on gce only. But in future we might need to change if other providers use kubemark too.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	// Copy all the logfiles that exist, to logDir.
+	for _, logfile := range logfiles {
+		logfileFullPath := filepath.Join(localLogPath, logfile + ".log*")  // Append .log* to copy rotated logs too.
+		cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("cp %v %v", logfileFullPath, logDir))
```
Why do we need "/bin/sh" here?

shyamjvs replied:
Because if we don't use bash, argument wouldn't be interpreted as regex (i.e. foobar.log* would be literally taken as the string "foobar.log*").

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+func uploadLogfilesToGCS(logDir string) error {
+	cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("ls %v/*", logDir))
+	if output, err := cmd.Output(); err != nil {
+		glog.Errorf("%v", )
```
remove

shyamjvs replied:
Oops! Sorry for the blooper.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+		glog.Errorf("%v", )
+		return fmt.Errorf("Could not list any logfiles: %v", err)
+	} else {
+		glog.Infof("List of logfiles available: %v", string(output))
```
Remove this else - this is useless.

shyamjvs replied:
This is the only place where the list of logfiles that are available on the node is printed. We wouldn't otherwise know directly from the container's logs of which files were attempted to be copied.
Do you still suggest removing this?

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	// Upload the files with compression and parallelism (speeds up on good network)
+  	// along with setting their ACL to make them publicly readable.
+  	gcsLogPath := filepath.Join(gcsArtifactsPath, nodeName)
+  	if ! strings.HasPrefix(gcsLogPath, "gs://") {
```
Can we just assume that the path is correct? i.e. validate it initially and assume that here it is correct?
Let's not try to make it too complicated.

shyamjvs replied:
Moved this logic to checkConfigValidity() function, as it seems more reasonable there. We cannot totally remove this part, because otherwise the exporter won't work even if we give a correct gcs path (as bash implicitly simplifies the argument "gs://foobar" to "gs:/foobar"). Other issues like path still not starting with "gs://", invalid path, etc would be caught by gsutil command and we don't need to add all those checking mechanisms.

shyamjvs replied:
Good news, got rid of the hack! It turns out that this issue of multiple slash simplification only occurs when we try calling the binary directly (not as part of container command), as then bash is in play. However in case of container too the same issue occurred but it was because of a different reason (which is filepath.Join() also does that simplification). Changed it now.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+		glog.Fatalf("Bad config provided: %v", err)
+	}
+
+	localTmpLogPath, err := ioutil.TempDir("/tmp", "k8s-systemd-logs")
```
Remove temp dir at the end (in defer).

shyamjvs replied:
Nice catch! Done.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	if err := uploadLogfilesToGCS(localTmpLogPath); err != nil {
+		glog.Fatalf("Could not upload logs to GCS: %v", err)
+	}
+	glog.Infof("Finished uploading logs successfully")
```
Logs successfully uploaded.

shyamjvs replied:
Done.

wojtek-t commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	}
+	glog.Infof("Finished uploading logs successfully")
+
+	if sleepDuration > 0 {
```
Remove if. Just leave:
```
time.Sleep(sleepDuration)
```

If it's 0, then it will just sleep for 0 time.

shyamjvs replied:
Done.

fejta commented on logexp/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+		glog.Infof("List of logfiles available: %v", string(output))
+	}
+
+	// Upload the files with compression and parallelism (speeds up on good network)
```
Might make more sense to note what -z and -m do closer to the gsutil command.

shyamjvs replied:
Done.

wojtek-t commented:
@fejta - just FYI, @shyamjvs just went for 2 weeks vacation, so this probably won't move during that time. But we will get back to it as soon as he is back.

shyamjvs commented:
@wojtek-t @fejta I'm back from vacation. I've addressed your comments, all except the one regarding fixing the 'sudo journalctl' command (explained the issue in [this](https://github.com/kubernetes/test-infra/pull/2310#discussion_r111983949) comment). While I try to find a fix for it, PTAL at the code now.

shyamjvs commented:
Fixed the remaining issue and tested the new image (details in [this](https://github.com/kubernetes/test-infra/pull/2310#discussion_r112054061) comment). PTAL.

fejta commented:
LGTM, will let Wojciech apply the label. Just click the merge button after that (or let me know that you want me to click it for you)

wojtek-t commented:
Sorry for delay - will take a look later today.
@fejta - thanks a lot for looking into it too.

wojtek-t commented:
/lgtm

Let's try this and iterate when we have something.
